internal/backend/runtime/omni/migration/migrations.go

```diff
@@ -1537,3 +1538,116 @@ func moveEtcdBackupStatuses(ctx context.Context, st state.State, logger *zap.Log
+	for clusterConfigVersion := range clusterConfigVersionList.All() {
+		clusterConfigVersionMap[clusterConfigVersion.Metadata().ID()] = clusterConfigVersion.TypedSpec().Value.Version
+	}
+
+	clusterMachineConfigList, err := safe.StateListAll[*omni.ClusterMachineConfig](ctx, st)
+	if err != nil {
+		return fmt.Errorf("failed to list cluster machine config patches: %w", err)
+	}
+
+	numNothingToDo, numDiskQuotaSupportEnabled, numApidCheckExtKeyUsageEnabled := 0, 0, 0
+
+	for clusterMachineConfig := range clusterMachineConfigList.All() {
+		id := clusterMachineConfig.Metadata().ID()
+		logger.Debug("checking cluster machine config patch", zap.String("id", id))
+
+		if clusterMachineConfig.Metadata().Phase() == resource.PhaseTearingDown {
+			continue
+		}
+
+		cluster, ok := clusterMachineConfig.Metadata().Labels().Get(omni.LabelCluster)
+		if !ok {
+			logger.Warn("cluster machine config does not have a cluster label, skipping...", zap.String("id", clusterMachineConfig.Metadata().ID()))
+
+			continue
+		}
+
+		clusterConfigVersion := clusterConfigVersionMap[cluster]
+		if clusterConfigVersion == "" {
+			logger.Warn("cluster does not have an initial version, skipping...", zap.String("id", clusterMachineConfig.Metadata().ID()), zap.String("cluster", cluster))
+
+			continue
+		}
+
+		initialVersion, err := semver.ParseTolerant(clusterConfigVersion)
+		if err != nil {
+			logger.Warn("failed to parse initial version, skipping...", zap.String("id", clusterMachineConfig.Metadata().ID()), zap.String("cluster", cluster), zap.Error(err))
+
+			continue
+		}
+
+		data, err := clusterMachineConfig.TypedSpec().Value.GetUncompressedData()
+		if err != nil {
+			return fmt.Errorf("failed to get uncompressed data: %w", err)
+		}
+
+		configDataStr := string(data.Data())
+
+		data.Free()
+
+		// see: https://github.com/siderolabs/omni/issues/1095
+		preserveDiskQuotaSupport := initialVersion.Major == 1 && initialVersion.Minor <= 4 && strings.Contains(configDataStr, "diskQuotaSupport: true")
+		preserveApidCheckExtKeyUsage := initialVersion.Major == 1 && initialVersion.Minor <= 2 && strings.Contains(configDataStr, "apidCheckExtKeyUsage: true")
+
+		if !preserveDiskQuotaSupport && !preserveApidCheckExtKeyUsage {
+			numNothingToDo++
+
+			continue
+		}
+
+		patchData := `machine:
+  features:
+`
+		if preserveDiskQuotaSupport {
+			patchData += "    diskQuotaSupport: true\n"
+
+			numDiskQuotaSupportEnabled++
+		}
+
+		if preserveApidCheckExtKeyUsage {
+			patchData += "    apidCheckExtKeyUsage: true\n"
+
+			numApidCheckExtKeyUsageEnabled++
+		}
+
+		configPatch := omni.NewConfigPatch(resources.DefaultNamespace, fmt.Sprintf("000-%s-preserve-version-contract", uuid.NewString()))
+
+		configPatch.Metadata().Labels().Set(omni.LabelCluster, cluster)
+		configPatch.Metadata().Labels().Set(omni.LabelClusterMachine, clusterMachineConfig.Metadata().ID())
+		configPatch.Metadata().Labels().Set(omni.LabelSystemPatch, "")
```

Comment:
@Unix4ever what do you think about this? I thought also about leaving it out, so it'd be more obvious to the users, and if they wanted, they could delete it on the UI.

Reply:
I don't know. It has downsides either way.
Maybe we can remove the patch and add meaningful name and description annotations on the resources.

It can even contain the URL to the issue in the description.

Reply:
changed it to do that.